include/deal.II/matrix_free/tools.h

```diff
@@ -27,7 +27,81 @@ DEAL_II_NAMESPACE_OPEN
  * evaluation.
  */
 namespace MatrixFreeTools
-{} // namespace MatrixFreeTools
+{
+  /**
+   * Modify @p additional_data so that cells are categorized
+   * according to their boundary IDs, making face integrals in the case of
+   * ECL simpler.
+   */
+  template <int dim, typename AdditionalData>
+  void
+  categorize_accoring_boundary_ids_for_ecl(
+    const Triangulation<dim> &tria,
+    AdditionalData &          additional_data,
+    const unsigned int        level = numbers::invalid_unsigned_int)
+  {
+    bool is_mg = (level != numbers::invalid_unsigned_int);
+
+    // ... create list for the category of each cell
+    if (is_mg)
+      additional_data.cell_vectorization_category.resize(
+        std::distance(tria.begin(level), tria.end(level)));
+    else
+      additional_data.cell_vectorization_category.resize(tria.n_active_cells());
+
+    // ... setup scaling factor
+    std::vector<unsigned int> factors(dim * 2);
+
+    std::map<unsigned int, unsigned int> bid_map;
+    for (unsigned int i = 0; i < tria.get_boundary_ids().size(); i++)
+      bid_map[tria.get_boundary_ids()[i]] = i + 1;
```

Comment:
Is it simpler if we write
```suggestion
    for (const auto id : tria.get_boundary_ids())
      bid_map[id] = bid_map.size() + 1;
```
(Not sure because it implicitly uses the `i+1` in this case.)